src/hazardous/base.rs

```diff
@@ -0,0 +1,156 @@
+/// Be careful if implementing this trait on your own. It cannot
+/// cause memory unsafety, and so is not marked `unsafe`. Implementing
+/// it can, however, lead to data types containing sensitive data ending
+/// up with APIs meant only for types containing only non-sensitive data.
+pub trait Public {}
+
+/// A small trait containing static information about the minimum and
+/// maximum size (in bytes) of a type containing data.
+pub trait Bounded {
+    /// The largest number of bytes this type should be allowed to hold.
+    const MIN: Option<usize> = None;
+
+    /// The smallest number of bytes this type should be allowed to hold.
+    const MAX: Option<usize> = None;
+}
+
+/// A generic holder for types that are basically just a bag of bytes
+/// with extra semantic meaning and restriction on top. We parameterize
+/// over the byte storage with parameter `B`. We parameterize over the
+/// API-level semantics of the type with phantom type `K`.
+#[derive(Clone)]
+pub struct Data<B, K> {
+    bytes: B,
+    phantom: PhantomData<K>,
+}
+
+impl<'a, B, K> Data<B, K>
+where
+    B: TryFrom<&'a [u8], Error = UnknownCryptoError>,
+    K: Bounded,
+{
+    /// TODO
+    pub fn from_slice(slice: &'a [u8]) -> Result<Self, B::Error> {
+        let min = K::MIN.unwrap_or(0);
+        let max = K::MAX.unwrap_or(usize::MAX);
+        if slice.len() < min || slice.len() > max {
+            return Err(UnknownCryptoError);
+        }
+
+        Ok(Self {
+            bytes: B::try_from(slice)?,
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<'a, B, K> Data<B, K>
+where
+    B: AsRef<[u8]>,
+{
+    /// Get the length of the contained byte slice.
+    pub fn len(&self) -> usize {
+        self.bytes.as_ref().len()
+    }
+
+    /// Check if the contained byte slice is empty.
+    pub fn is_empty(&self) -> bool {
+        self.bytes.as_ref().is_empty()
+    }
+}
+
+impl<'a, B, K> AsRef<[u8]> for Data<B, K>
+where
+    B: AsRef<[u8]>,
+    K: Public,
+{
+    /// Get a reference to the underlying byte slice.
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.as_ref()
+    }
+}
+
+impl<'a, B, K> Data<B, K>
+where
+    B: AsRef<[u8]>,
+    K: Secret,
+{
+    /// TODO: Grab docs for `unprotected_as_bytes` and insert here.
+    pub fn unprotected_as_bytes(&self) -> &[u8] {
+        self.bytes.as_ref()
+    }
+}
+
+// We implement this manually to skip over the PhantomData.
+impl<B, K> PartialEq for Data<B, K>
+where
+    B: PartialEq<B>,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.bytes.eq(&other.bytes)
+    }
+}
+
+// We implement this manually to skip over the PhantomData.
+impl<B, K> fmt::Debug for Data<B, K>
+where
+    B: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.bytes.fmt(f)
+    }
+}
+
+/// A convenient type for holding data with a static upper bound on
+/// its size. The bytes are held with a static array.
+#[derive(Clone, Debug)]
```

Comment:
We shouldn't ever derive debug or clone for any data-covering type if we don't know whether it holds secret or public data, is my first thought here.

Reply:
Agreed. I believe I caught and fixed this in the next iteration I'm about to publish.